### PR TITLE
Use POST Instead of GET for API Token Generation

### PR DESCRIPTION
**Pull request by jbarnett-r7**
This PR updates the `/api/v1/auth/generate-token` URL to use POST instead of GET requests. The `username` and `password` fields will now be passed as JSON in the body of the request. It also updates the `msfdb` script and test page at `/api/v1/auth/login` to use the new method.

## Verification

List the steps needed to make sure this thing works

- [x] Run `msfdb init` to set up a metasploit API webservice and database. You should be successfully given a username, password, and valid API token.
- [x] Navigate to `/api/v1/api-docs` and test that the `generate-token` endpoint is successful when inputting the username and password from the `msfdb` setup
- [x] Verify the token you receive is valid by loading up `msfconsole` and configuring a data service connection `db_connect -t <token you were given> --skip-verify https://localhost:8080`
- [x] Navigate to `/api/v1/auth/login`. Enter your username and password and verify you are given the current API token back. Also test the `Generate Net Token` link in the upper left corner to ensure you get a new token each time.

### CODE REVIEW

File: lib/msf/core/web_services/views/auth/login.erb [hunk 1]
```diff
@@ -7,20 +7,53 @@
 </head>
 <body>
 
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script>
```
Comment (bcoles) [inline lib/msf/core/web_services/views/auth/login.erb]: Third-party hosted outdated jQuery without SRI.

outdated jQuery

SRI

File: documentation/api/v1/auth_api_doc.rb [hunk 2]
```diff
@@ -16,25 +16,20 @@ module AuthApiDoc
 
   swagger_path '/api/v1/auth/generate-token' do
     # Swagger documentation for /api/v1/auth/generate-token GET
```
Comment (mkienow-r7) [inline documentation/api/v1/auth_api_doc.rb]: Minor: comment should state POST instead of GET

File: lib/msf/core/web_services/authentication/strategies/user_password.rb [hunk 3]
```diff
@@ -16,15 +16,19 @@ class UserPassword < Warden::Strategies::Base
       # Check if request contains valid data and should be authenticated.
       # @return [Boolean] true if strategy should be run for the request; otherwise, false.
       def valid?
-        params['username'] && params['password']
+        body = JSON.parse(request.body.read, symbolize_names: true)
```
Comment (mkienow-r7) [inline lib/msf/core/web_services/authentication/strategies/user_password.rb]: Should this module `require 'json'` should it be used somewhere JSON has not already been required?

File: lib/msf/core/web_services/authentication/strategies/user_password.rb [hunk 4]
```diff
@@ -16,15 +16,19 @@ class UserPassword < Warden::Strategies::Base
...
+        request.body.rewind # Reset the StringIO buffer so any further consumers can read the body
```
Comment (mkienow-r7) [inline lib/msf/core/web_services/authentication/strategies/user_password.rb]: Would it be a good idea to have a begin-ensure block around the operations of the method and move the `request.body.rewind` to the `ensure` as a fail-safe should the JSON received cause an error on `JSON.parse`?
```
begin
  body = JSON.parse(request.body.read, symbolize_names: true)
  body[:username] && body[:password]
ensure
  request.body.rewind # Reset the StringIO buffer so any further consumers can read the body
end
```

File: lib/msf/core/web_services/views/auth/login.erb [hunk 5]
```diff
@@ -7,20 +7,56 @@
 </head>
 <body>
 
+<script
+  src="https://code.jquery.com/jquery-3.3.1.min.js"
```
Comment (mkienow-r7) [inline lib/msf/core/web_services/views/auth/login.erb]: If the user is on an isolated network without internet access then the login form will not operate, correct?

Comment (jbarnett-r7) [reply lib/msf/core/web_services/views/auth/login.erb]: This is true. The other options are to include jquery in the repo or to do straight Javascript AJAX. This login form is not really a primary feature at this point, so I'm not sure that it's worth it to write our own yet. When we re-visit this and make it a first class feature we should definitely remove the jquery requirement.

File: lib/msf/core/web_services/views/auth/login.erb [hunk 6]
```diff
@@ -7,20 +7,56 @@
...
     <div class="credential-container">
       <h2>Log In - Metasploit API</h2>
       <label for="username"><b>Username</b></label>
       <input type="text" placeholder="Enter Username" name="username" required>
 
       <label for="password"><b>Password</b></label>
       <input type="password" placeholder="Enter Password" name="password" required>
-
-      <button type="submit">Log In</button>
+      <input id="submit" type="button" name="submit" value="submit">
```
Comment (mkienow-r7) [inline lib/msf/core/web_services/views/auth/login.erb]: Minor: the button text was changed from "Log In"

File: lib/msf/core/web_services/views/auth/login.erb [hunk 7]
```diff
@@ -7,20 +7,56 @@
...
+            json[this.name] = this.value || '';
+        });
+
+        return JSON.stringify(json);
+    }
+
+    $(document).ready(function(){
+        $("#submit").on('click', function(){
+            $.ajax({
+                url: '<%= AuthServlet.api_generate_token_path %>',
```
Comment (mkienow-r7) [inline lib/msf/core/web_services/views/auth/login.erb]: This should be `AuthServlet.api_login_path`, otherwise, the act of the user logging into the system generates a new token.

**`msfdb reinit` output:**
```
MSF web service user API token: 23474c55a38a3db370745e0a7c3f64a7f3271100c203a7795a407d2a761aec2dc18ed2dd4af79017
```

**API account display after logging in:**
```
Current API Token: bfea307a9be82456a00a661c4e0cfd850d0edec796ff042538b7e34ec19620ceedcb02f579a3dee5
```

File: lib/msf/core/web_services/views/auth/login.erb (hunk 6 above)
Comment (mkienow-r7) [reply lib/msf/core/web_services/views/auth/login.erb]: In addition, `lib/msf/core/web_services/public/msf-ws.css` contains styling for the button so that it filled the bottom of the log on credential container `div`. Update that selector so that it applies to this new input button.

File: lib/msf/core/web_services/views/auth/login.erb (hunk 7 above)
Comment (mkienow-r7) [reply lib/msf/core/web_services/views/auth/login.erb]: Regarding the associated verification step:
> Navigate to `/api/v1/auth/login`. Enter your username and password and verify you are given a valid token back. Also test the `Generate Net Token` link in the upper left corner to ensure you get a new token each time.

Logging into the API account should not generate a new API token. When the user logs in they should be viewing the current API token. Clicking the `Generate New API Token` link would generate a new API token for the user.

Comment (jbarnett-r7) [reply lib/msf/core/web_services/views/auth/login.erb]: We discussed this in depth privately, but just adding notes here:

I attempted to make this change, but ran into issues because the POST `/login` endpoint only returns 30x redirects from the server, whether the login request is valid or not. Javascript AJAX requests don't handle this well because as far as it's concerned both responses mean the same thing. 

We need to make some bigger changes and discussions around how we want to handle the login functionality. One of the key requests by @mkienow-r7 is to ensure the server has knowledge of the origin page that sent the user to the `/login` page so we can redirect them there after the login is successful. Implementing these changes is outside of the scope of this PR and will be made in the future.

As a stop-gap, the `/auth/login` page will always redirect to `/auth/account` on both valid and invalid login attempts. I did add some text to the `/auth/account` page to make it more apparent to the user that they are not currently logged in and redirect them back to the login page to try again.

**Comment by jbarnett-r7** [comment]
@msjenkins-r7 test this please

File: lib/msf/core/web_services/views/auth/login.erb (hunk 7 above)
Comment (mkienow-r7) [reply lib/msf/core/web_services/views/auth/login.erb]: Thanks for making the note and created associated tickets @jbarnett-r7!

**Comment by mkienow-r7** [comment]
@msjenkins-r7 test this please

**Comment by mkienow-r7** [comment]
# Release Notes

This changes the `UserPassword` authentication to receive the username and password fields in the JSON request body rather than the URL query string. The `/api/v1/auth/generate-token` endpoint was updated to use the POST request method instead of GET. This required associated updates to use the new request method in the `msfdb` script and API account login page (`/api/v1/auth/login`). It is important to note that if the user is on an isolated network without internet access then the login form will not operate due to it pulling in jQuery from the internet.
